### PR TITLE
[WIP] JWT revamp

### PR DESCRIPTION
**Pull request by williamchong007**
Done:
+ Add AUD ISS and permissions field
+ Add RSA support
+ Add jwtId for session control

TODO in future:
+ Add void of session

### CODE REVIEW

**Comment by williamchong007** [comment]
`setSessionCookie()` seems very questionable now, maybe need to discuss

File: server/api/users.js [hunk 1]
```diff
@@ -81,15 +81,21 @@ function setSessionCookie(req, res, token) {
...
     wallet,
     permissions: ['read', 'write', 'like'],
   };
-  const token = jwtSign(payload);
+  const { token, jwtid } = jwtSign(payload);
   res.cookie('likecoin_auth', token, AUTH_COOKIE_OPTION);
   setSessionCookie(req, res, token);
+  await dbRef.doc(user).collection('session').doc(jwtid).create({
+    lastAccessedUserAgent: req.headers['user-agent'] || 'unknown',
+    lastAccessedIP: req ? (req.headers['x-real-ip'] || req.ip) : undefined,
```
Comment (nnkken) [inline server/api/users.js]: So it is possible to have falsy value `req`? Then `setSessionCookie` should care about this?

File: server/api/users.js [hunk 2]
```diff
@@ -320,14 +326,19 @@ router.put('/users/new', apiLimiter, multer.single('avatar'), async (req, res, n
...
   const { wallet } = req.body;
   if (req.user.wallet !== wallet) {
     res.status(401).send('LOGIN_NEEDED');
     return;
   }
   setSessionCookie(req, res, req.cookies.likecoin_auth);
   res.sendStatus(200);
+  await dbRef.doc(req.user.user).collection('session').doc(req.user.jti).update({
+    lastAccessedUserAgent: req.headers['user-agent'] || 'unknown',
+    lastAccessedIP: req ? (req.headers['x-real-ip'] || req.ip) : undefined,
```
Comment (nnkken) [inline server/api/users.js]: Same as line 95

File: server/api/users.js (hunk 1 above)
Comment (williamchong007) [reply server/api/users.js]: actually not possible here, copy and paste programming fault..

**Pull request merged**